File: SVM/src/svm.py
#!/usr/bin/python3
import pandas as pd
import numpy as np

from sklearn.svm import LinearSVC
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV

import os
THIS_FOLDER = os.path.dirname(os.path.abspath(__file__))
data_path = THIS_FOLDER + '/../../Features/features.csv'
data = pd.read_csv(data_path, header=None)
X = data.loc[:, :1023]
y = data[1024]
filename = data[1025]

X_train, X_test, y_train, y_test, filename_train, filename_test = train_test_split(X, y, filename, test_size=0.33, random_state=42)

label_unique = np.unique(y)
train_label_unique = np.unique(y_train)
test_label_unique = np.unique(y_test)

print(f"Train set contains {len(train_label_unique)} labels")
print(f"Train set contains {X_train.shape[0]} features")
print(f"Test set contains {len(test_label_unique)} labels")
print(f"Test set contains {X_test.shape[0]} examples")

clf = make_pipeline(StandardScaler(), LinearSVC(random_state=0))
clf.fit(X_train, y_train)

scaler = clf.named_steps['standardscaler']
scale = scaler.scale_
mean = scaler.mean_

svc = clf.named_steps['linearsvc']
coef = svc.coef_
intercept = svc.intercept_

print("mean: ", mean)
print("scale: ", scale)
print("coef: ", coef)
print("intercept: ", intercept)



score = clf.score(X_test, y_test)
print("score: ",score)


# Save the parameters to a file
print("- saving npy parameters -")
# np.save('scale.npy', scale)
# np.save('mean.npy', mean)
# np.save('coef.npy', coef)
# np.save('intercept.npy', intercept)
np.savetxt('scale.txt', scale)
np.savetxt('mean.txt', mean)
np.savetxt('coef.txt', coef)
np.savetxt('intercept.txt', intercept)

# scale_load = np.load('scale.npy')
# for i in range(scale_load.shape[0]):
#     print(scale_load[i])

File: SVM/src/svm.cpp
#include <vector>
#include <iostream>
#include "load_utils.h"

int main()
{
  // Load parameters
  std::vector<std::vector<double>> coef = load_npy_txt_2d("../src/storage/coef.txt");
  std::vector<double> scale = load_npy_txt("../src/storage/scale.txt");
  std::vector<double> mean = load_npy_txt("../src/storage/mean.txt");
  std::vector<double> intercept = load_npy_txt("../src/storage/intercept.txt");

  // Load and normalize features
  std::vector<std::vector<double>> features;
  std::vector<int> labels;

  load_test_set(features, labels);
  normalize(features, mean, scale);

  // Predict the class labels with an svm
  std::cout << "features.size(): " << features.size() << '\n';
  std::cout << "coef.size(): " << coef.size() << '\n';

  std::vector<int> predictions(features.size());
  float percent = 0;
  for (size_t i = 0; i < features.size(); i++) {
    double max_score = -1e18;
    int max_idx = -1;
    for (size_t j = 0; j < coef.size(); j++) {
      double score = 0;
      for (size_t k = 0; k < 1024 ; k++) {
        score += coef[j][k] * features[i][k];
      }
      score += intercept[j];
      if (score > max_score) {
        max_score = score;
        max_idx = j;
      }
    }
    predictions[i] = max_idx;
    if (predictions[i] == labels[i]){percent++;}
  }

  std::cout << "We found : "<< percent/features.size()*100. << "%" << '\n';

  return 0;
}
